src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -62,111 +61,115 @@
+     (log/info "Borrowed all JRuby instances, proceeding with cleanup.")
+     (send-agent (get-modify-instance-agent pool-context)
+                 #(cleanup-and-refill-pool pool-context old-instances refill? on-complete)))))
 
 (schema/defn ^:always-validate
   flush-pool-for-shutdown!
   "Flush of the current JRuby pool when shutting down during a stop.
   Delivers the on-complete promise when the pool has been flushed."
-  ;; Since this function is only called by the pool-agent, we know that if we
+  ;; Since the drain-pool! function takes the pool lock, we know that if we
   ;; receive multiple flush requests before the first one finishes, they will
-  ;; be queued up and we don't need to worry about race conditions between the
-  ;; steps we perform here in the body.
+  ;; be queued up waiting for the lock, which will never be granted because this
+  ;; function does not refill the pool, but instead inserts a shutdown poison pill
   [pool-context :- jruby-schemas/PoolContext
    on-complete :- IDeref]
-  (try
-    (log/info "Flush request received; creating new JRuby pool.")
-    (let [{:keys [config]} pool-context
-          new-pool-state (jruby-internal/create-pool-from-config config)
-          new-pool (:pool new-pool-state)
-          old-pool-state (jruby-internal/get-pool-state pool-context)
-          old-pool (:pool old-pool-state)
-          old-pool-size (:size old-pool-state)]
-      (when-not (pool-initialized? old-pool-size old-pool)
-        (throw (IllegalStateException. "Attempting to flush a pool that does not appear to have successfully initialized. Aborting.")))
-      (.insertPill new-pool (ShutdownPoisonPill. new-pool))
-      (swap-and-drain-pool! pool-context old-pool-state new-pool-state false))
-    (finally
-      (deliver on-complete true))))
-
-(schema/defn ^:always-validate
-  flush-and-repopulate-pool!
-  "Flush of the current JRuby pool.  NOTE: this function should never
-  be called except by the pool-agent."
-  [pool-context :- jruby-schemas/PoolContext]
-  ;; Since this function is only called by the pool-agent, we know that if we
-  ;; receive multiple flush requests before the first one finishes, they will
-  ;; be queued up and we don't need to worry about race conditions between the
-  ;; steps we perform here in the body.
-  (log/info "Flush request received; creating new JRuby pool.")
-  (let [{:keys [config]} pool-context
-        new-pool-state (jruby-internal/create-pool-from-config config)
-        old-pool (jruby-internal/get-pool-state pool-context)]
-    (swap-and-drain-pool! pool-context old-pool new-pool-state true)))
+  (log/debug "Beginning flush of JRuby pools for shutdown")
+  (let [pool-state (jruby-internal/get-pool-state pool-context)
+        pool (:pool pool-state)]
+    (drain-and-refill-pool! pool-context false on-complete)
+    (jruby-internal/insert-shutdown-poison-pill pool))
+  (log/debug "Finished flush of JRuby pools for shutdown")
```

Comment:
Seems like the "Finished flush" log statement should be after the `@on-complete` one?